sklearn/linear_model/ridge.py

```diff
@@ -367,11 +367,28 @@ def _ridge_regression(X, y, alpha, sample_weight=None, solver='auto',
                       return_n_iter=False, return_intercept=False,
                       X_scale=None, X_offset=None):
 
-    if return_intercept and sparse.issparse(X) and solver != 'sag':
-        if solver != 'auto':
-            warnings.warn("In Ridge, only 'sag' solver can currently fit the "
-                          "intercept when X is sparse. Solver has been "
-                          "automatically changed into 'sag'.")
+    has_sw = sample_weight is not None
+
+    if solver == 'auto':
+        if return_intercept:
+            # only sag supports fitting intercept directly
+            solver = "sag"
+        elif has_sw:
+            # this should be changed since all solvers support sample_weights
+            solver = "cholesky"
+        elif sparse.issparse(X):
+            solver = "sparse_cg"
```

Comment:
I find the sequence of conditions hard to follow. I find the initial pattern easier:
```python
if return_intercept:
    solver = "sag"
elif not sparse.issparse(X) or has_sw:
    solver = "cholesky"
else:
    solver = "sparse_cg"
```

If all solvers support sample weight, I'd be in favor of removing the `has_sw` condition.

Reply:
ok, I changed the sequence and removed the `has_sw` option.